feat(conditional-layers): Implement generalized version of tri-layer support for "adjust" layer

bcat opened the pull request:
I missed tri-layer support from QMK, so I implemented a ZMK version. It seemed most logical to express as an N-to-1 mapping of `if-layers` to activate a single `then-layer`, so I called it "conditional layers".

As discussed in the attached issue, I considered calling this "layer combos" and replacing the `then-layer` property with a `bindings` property that could trigger an `&mo` behavior (or any other behavior as well), but in practice I didn't see any reason to generalize that much.

Fixes #828.

### Tasks before review

- [X] Initial implementation
- [X] Basic unit tests
- [x] Clarity on two TODOs in the code (see linked issue for more details)
- [x] Documentation
- [x] Unit tests for different orderings of layer down/up
- [x] Unit tests for multiple conditional layer configs in the same keymap
- [x] Testing on real keyboard

caksoylar commented:
Hey @bcat, I was missing tristate layers from my QMK config thinking about implementing myself until I found this. This PR works pretty well for me at the moment and it would be great to have this available in ZMK. You bring up some good points for discussion in https://github.com/zmkfirmware/zmk/issues/828 and the TODO items in the code, but IMO it would still be a very good addition with all the current caveats (i.e. do not define chains of multi-layers, do not activate the then layer through multiple methods simultaneously); I am not sure whether/how QMK handles those issues either.

PS: I kind of like the `conditional layers` name better than `multi layers` as well; it fits the if-then terminology too.

bcat commented:
Thanks for testing it! I sort of let this stagnate both because I wasn't sure if anyone else had a use for it and because life has just been busy lately. But I may have some time this weekend to update the PR (e.g., finish up the unit tests) and put it out for a real review.

I do agree on "conditional layers" as a clearer name. That should be easy enough to change. And I agree the edge cases are probably fine for now. The big open question for me is the second point on https://github.com/zmkfirmware/zmk/issues/828#issuecomment-860344899. Namely should the "then layer" just use a normal binding specification (like combos do). That seems a bit cleaner to me. If I have some cycles this weekend I might give that a try.

caksoylar commented:
> Namely should the "then layer" just use a normal binding specification (like combos do). 

I don't have much of a preference either way, but I can't think of a useful behavior other than `&mo` like you said in your comment. If that is the case, I would tend towards keeping it simple as it is.

bcat commented:
> > Namely should the "then layer" just use a normal binding specification (like combos do).
> 
> I don't have much of a preference either way, but I can't think of a useful behavior other than `&mo` like you said in your comment. If that is the case, I would tend towards keeping it simple as it is.

Yeah, I think `&mo` is the only practically useful binding there. IIRC, I mostly liked the idea of using bindings because it sidestepped the problem of imbalanced layer activate/deactive actions (since the actual manipulation of the layer state was done by the standard, existing behavior). But honestly I think it depends if it makes the implementation more complex or not. I'll have to take a closer look at the code and see.

bcat commented:
Rebased against latest main, renamed to "conditional layers", and completed docs/tests. Tomorrow I'll make sure this still works on a physical keyboard, and then I'll send it out for review.

bcat commented:
Seems to still work fine on my Crkbd (with nice!nano v1). Should be ready for review now. :)

@caksoylar, mind giving the latest version of this PR a try when you have a chance and verifying I haven't broken anything for you? Thanks!

bcat commented:
Here's an example keymap that uses this feature: https://github.com/bcat/zmk-config/commit/63f4dbb61c9d5fd8ff94273b09965d63b8fd6590

caksoylar commented:
I updated my fork with the PR, then [readjusted my keymap](https://github.com/caksoylar/zmk-config/commit/e19e1df4b914441151c961a6be136506655880c9) for the latest changes. Still working great here, thanks for your work!

caksoylar commented on docs/docs/features/conditional-layers.md:
```diff
@@ -0,0 +1,51 @@
+
+:::info
+One conditional layer configuration can trigger another, possibly many times in succession; however,
+this process must eventually end since there are only a finite number of layers.
```
I had some trouble parsing this statement. Does this mean that an infinite loop isn't technically possible? I assume so, if `if-layers` is constrained to have at least two elements.

bcat replied:
Yeah, as far as a I know, an infinite loop shouldn't be possible since the implementation only tries to activate a layer if it's not already active, and so in the worst case you end up done when every layer is active. But honestly, on further thought I don't think that detail is really important, so I pruned down the note text. WDYT?

caksoylar replied:
I think that's good, the user doesn't really need to know whether it can cause an issue anyway, especially if it cannot :)

caksoylar commented on docs/docs/features/conditional-layers.md:
```diff
@@ -0,0 +1,51 @@
+title: Conditional Layers
+---
+
+Conditional layers support activates a particular layer (called the then-layer) when all layers in a
```
```suggestion
Conditional layers support activates a particular layer (called the `then-layer`) when all layers in a
```

caksoylar commented on docs/docs/features/conditional-layers.md:
```diff
@@ -0,0 +1,51 @@
+---
+
+Conditional layers support activates a particular layer (called the then-layer) when all layers in a
+specified set (called the if-layers) are active. This feature generalizes what's commonly known as
```
```suggestion
specified set (called the `if-layers`) are active. This feature generalizes what's commonly known as
```

caksoylar commented on docs/docs/features/conditional-layers.md:
```diff
@@ -0,0 +1,51 @@
+title: Conditional Layers
+---
+
+Conditional layers support activates a particular layer (called the `then-layer`) when all layers in
```
```suggestion
Conditional layers support activating a particular layer (called the `then-layer`) when all layers in
```

caksoylar commented on docs/docs/features/conditional-layers.md:
````diff
@@ -0,0 +1,51 @@
+};
+```
+
+Each conditional layer configuration may have whatever name you like, but it's helpful to chose
````
```suggestion
Each conditional layer configuration may have whatever name you like, but it's helpful to choose
```

caksoylar commented on docs/docs/features/conditional-layers.md:
```diff
@@ -0,0 +1,51 @@
+Each conditional layer configuration may have whatever name you like, but it's helpful to chose
+something self explanatory like `tri_layer`. The following properties are supported:
+
+- The `if-layers` property specifies a set of layer numbers, all of which must be active for the
```
```suggestion
- `if-layers` specifies a set of layer numbers, all of which must be active for the
```

caksoylar commented on docs/docs/features/conditional-layers.md:
```diff
@@ -0,0 +1,51 @@
+
+- The `if-layers` property specifies a set of layer numbers, all of which must be active for the
+  conditional layer to trigger.
+- The `then-layer` property specifies a layer number that should be activated if and only if all the
```
```suggestion
- `then-layer` specifies a layer number that should be activated if and only if all the
```

caksoylar commented on docs/docs/features/conditional-layers.md:
```diff
@@ -0,0 +1,51 @@
+- The `then-layer` property specifies a layer number that should be activated if and only if all the
+  layers specified in the `if-layers` property are active.
+
+Therefore, in this example, layer 3 ("adjust") will be activated if and only if both of layers 1
```
```suggestion
Therefore, in this example, layer 3 ("adjust") will be activated if and only if both layers 1
```

caksoylar commented on docs/docs/features/conditional-layers.md:
```diff
@@ -0,0 +1,51 @@
+layer](../behaviors/layers.md#momentary-layer) behavior), it will be immediately deactivated if the
+associated `then-layers` configuration is not met. As such, we recommend avoiding using regular
+layer behaviors for `then-layer` targets.
+:::
```
Can we add a note similar to the point 2. in [QMK docs](https://docs.qmk.fm/#/ref_functions?id=update_tri_layer_statestate-x-y-z), saying that the `then-layer` should be higher than the `if-layers` to be able to access it reliably?

caksoylar reviewed (commented):
I noted a few language improvements that I missed the first time around, and suggestion to add a remark regarding layer order in the docs.

bcat commented:
Thanks! I've incorporated those suggestions as well.

BTW, I'm not used to reviews cycles in PRs. Instead of accumulating lots of tiny commits, should I squash them into more logical chunks and force push? I'm happy to do whatever is preferred. (I know QMK maintainers often squash the whole PR into one commit when it's merged, so it's a moot point, but I wasn't sure if ZMK did the same.)

caksoylar reviewed (approved):
Thanks for incorporating the suggestions! I can approve in terms of docs, but other contributors will need to review. 

Regarding commits, AFAIK PRs can be merged as is as long as individual commits make sense, but they can get squashed otherwise. If you feel it can be improved you can reorganize/rebase them but I wouldn't worry about it too much.

caksoylar commented:
I also haven't had any issues since I started using it last week.

petejohanson reviewed (approved):
This is perfect! Can you rebase to fix the merge conflict so we can merge?

Thanks!

bcat commented:
> This is perfect! Can you rebase to fix the merge conflict so we can merge?

Thanks for the review! Rebased (and squashed some of doc revision commits while I was in there) and pushed. All the tests still pass, though as mentioned on Discord, I don't currently have a presently have a working ZMK board available to test on real hardware.

bcat commented:
> This is perfect! Can you rebase to fix the merge conflict so we can merge?

Thanks for the review! Rebased (and squashed some of doc revision commits while I was in there) and pushed. All the tests still pass, though as mentioned on Discord, I don't currently have a presently have a working ZMK board available to test on real hardware.

caksoylar commented:
I have been rebasing this on `main` and using it for the last the couple of months, I can confirm that the newly pushed version works fine as well.

bcat commented:
Of course I managed to screw up a one-line merge. :) Fixed `sidebar.js` now, for realz.

jordiorlando commented:
I'm thrilled that this has been merged, but I'm seeing an issue in a corner case. [My keymap](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) has layer-tap space keys for `raise` and `lower`, and I can't get the `adjust` layer to activate with this configuration. Any advice?

bcat commented:
> I'm thrilled that this has been merged, but I'm seeing an issue in a corner case. [My keymap](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) has layer-tap space keys for `RAISE` and `LOWER`, and I can't get the `CONFIG` layer to activate with this configuration. Any advice?

Oh, that is interesting. The intent of the new functionality was that it would work with any type of layer binding (since the implementation just listens for layer state changes rather than monitoring individual behaviors). But I can definitely believe that I missed some edge case. :)

The next few days are likely to be busy for me, unfortunately, but I can try to isolate a unit test that repros the LT bug (if nobody else beats me to it).

bcat commented:
So I did the simplest, quickest thing and changed https://github.com/zmkfirmware/zmk/blob/30ed2375335acf47da0216c6832ecfd30ea7097c/app/tests/conditional-layer/tri-layer/native_posix.keymap locally to use `&lt` instead of `&mo` (lengthening the `ZMK_MOCK_PRESS` durations to ensure the "hold" triggers). But that modified test still passes. So it's not as simple as "layer taps don't work with conditional layers". @jordiorlando, I think there's some more complex interaction with different parts of your keymap going on here. I'm not sure exactly what yet, though. Will need to dig a bit...

jcmkk3 commented:
> I'm thrilled that this has been merged, but I'm seeing an issue in a corner case. [My keymap](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) has layer-tap space keys for `RAISE` and `LOWER`, and I can't get the `CONFIG` layer to activate with this configuration. Any advice?

Have you tried to switch your LT to be MO just to ensure that the conditional-layer functionality is working for you?

jordiorlando commented:
> > I'm thrilled that this has been merged, but I'm seeing an issue in a corner case. [My keymap](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) has layer-tap space keys for `RAISE` and `LOWER`, and I can't get the `CONFIG` layer to activate with this configuration. Any advice?
> 
> Have you tried to switch your LT to be MO just to ensure that the conditional-layer functionality is working for you?

Good idea, it's possible my build isn't pulling the latest version of ZMK. Trying that now.

bcat commented:
Indeed, I think that is the issue. The `west.yml` file for this config points at a third-party fork of ZMK: https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/a521a785453668adba0678df668b81436cd29532/config/west.yml. Specifically, this fork: https://github.com/LOWPROKB/zmk. Which seems a bit outdated (last commit there is e6645882c32454251466949fa9a6447b195acd72, before the conditional layers PR was merged).

Is this keyboard supported in upstream ZMK? If so, you can probably just switch your config repo to upstream ZMK. But if not, sadly I think you'll need to create your own fork of the third-party repo and manually integrate this PR for the time being.

@petejohanson, it looks like the "Corne-ish Zen" maker's sample repo points at a nonstandard ZMK fork (https://github.com/LOWPROKB/zmk-config-Corne-ish-Zen/blob/b8de4f2b6545712141cc4f4c19260eaae89d2245/config/west.yml), so I suspect lots of folks are copying that by default. :( Is that a supported practice? I honestly don't mean to stir up drama (and will happily shut up if this is an existing issue), but it seems like this may lead to lots of end users building off ZMK forks with various degrees of bitrot....

bcat commented:
Somewhat separately... I'm surprised the build doesn't just fail if there's unknown stuff in the keymap file. I don't really know how devicetree works under the hood, but is there some way to guard against this at build time rather than just silently ignoring the unsupported bits of the keymap?

jordiorlando commented:
This would be a good question for @LOWPROKB as well. I think there was talk of upstreaming support for the Corne-ish Zen, but I don't know the current status.

bcat commented:
Indeed, that's fair. Also, as it looks like there is not a bug with the conditional layers feature itself, maybe we should move the discussion about upstreaming, forks, etc. to a new issue? (Not trying to be too pedantic, it just seems like the root cause is not related.)

jordiorlando commented:
> Indeed, that's fair. Also, as it looks like there is not a bug with the conditional layers feature itself, maybe we should move the discussion about upstreaming, forks, etc. to a new issue? (Not trying to be too pedantic, it just seems like the root cause is not related.) But then again, I don't really know _where_ a better to place to discuss would be....

Agreed, I don't want to flood this thread with off-topic jabber. Thank you for the super quick help troubleshooting!

I created a new issue over [here](https://github.com/LOWPROKB/zmk-config-Corne-ish-Zen/issues/3) that is probably more suitable for discussion.

jordiorlando commented:
Just wanted to leave a final comment to say that I've tested [my layout](https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/main/config/corne-ish_zen.keymap) with the latest version of zmk, and everything works as intended. So stoked to have this feature! Thanks all.

petejohanson commented:
> Indeed, I think that is the issue. The `west.yml` file for this config points at a third-party fork of ZMK: https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/a521a785453668adba0678df668b81436cd29532/config/west.yml. Specifically, this fork: https://github.com/LOWPROKB/zmk. Which seems a bit outdated (last commit there is [e664588](https://github.com/zmkfirmware/zmk/commit/e6645882c32454251466949fa9a6447b195acd72), before the conditional layers PR was merged).
> 
> Is this keyboard supported in upstream ZMK? If so, you can probably just switch your config repo to upstream ZMK. But if not, sadly I think you'll need to create your own fork of the third-party repo and manually integrate this PR for the time being.
> 
> @petejohanson, it looks like the "Corne-ish Zen" maker's sample repo currently points at a nonstandard ZMK fork (https://github.com/LOWPROKB/zmk-config-Corne-ish-Zen/blob/b8de4f2b6545712141cc4f4c19260eaae89d2245/config/west.yml), so I suspect lots of folks are copying that by default. :( Is that a supported practice? I honestly don't mean to stir up drama (and will happily shut up if this is already a known issue), but it seems like this may lead to lots of end users building off ZMK forks with various degrees of bitrot....

This is an unfortunate situation, certainly, and something that the modular work should help formalize. The battery changes in particular are the main holdup.

bcat commented:
> > Indeed, I think that is the issue. The `west.yml` file for this config points at a third-party fork of ZMK: https://github.com/jordiorlando/zmk-config-Corne-ish-Zen/blob/a521a785453668adba0678df668b81436cd29532/config/west.yml. Specifically, this fork: https://github.com/LOWPROKB/zmk. Which seems a bit outdated (last commit there is [e664588](https://github.com/zmkfirmware/zmk/commit/e6645882c32454251466949fa9a6447b195acd72), before the conditional layers PR was merged).
> > Is this keyboard supported in upstream ZMK? If so, you can probably just switch your config repo to upstream ZMK. But if not, sadly I think you'll need to create your own fork of the third-party repo and manually integrate this PR for the time being.
> > @petejohanson, it looks like the "Corne-ish Zen" maker's sample repo currently points at a nonstandard ZMK fork (https://github.com/LOWPROKB/zmk-config-Corne-ish-Zen/blob/b8de4f2b6545712141cc4f4c19260eaae89d2245/config/west.yml), so I suspect lots of folks are copying that by default. :( Is that a supported practice? I honestly don't mean to stir up drama (and will happily shut up if this is already a known issue), but it seems like this may lead to lots of end users building off ZMK forks with various degrees of bitrot....
> 
> This is an unfortunate situation, certainly, and something that the modular work should help formalize. The battery changes in particular are the main holdup.

Makes sense. I guess the only thing I'd like to find out then is whether it's expected that there was no build failure here. Maybe that's just how devicetree works, but it kind of surprises me.

petejohanson commented:
> 
> Makes sense. I guess the only thing I'd like to find out then is whether it's expected that there was no build failure here. Maybe that's just how devicetree works, but it kind of surprises me.

Yeah, unfortunately, the current build doesn't prevent this. I do have some ideas for how to improve this.
